ui/components/app/toast-master/toast-master-selectors.ts

```diff
@@ -0,0 +1,118 @@
+ *
+ * @param state - The application state containing the necessary survey data.
+ * @returns True if the current time is between the survey start and end times and the survey link was not last clicked or closed. False otherwise.
+ */
+export function getShowSurveyToast(state: State): boolean {
+  if (state.metamask?.surveyLinkLastClickedOrClosed) {
+    return false;
+  }
+
+  const startTime = new Date(`${SURVEY_DATE} ${SURVEY_START_TIME}`).getTime();
+  const endTime = new Date(`${SURVEY_DATE} ${SURVEY_END_TIME}`).getTime();
+  const now = Date.now();
+
+  return now > startTime && now < endTime;
+}
+
+/**
+ * Determines if the privacy policy toast should be shown based on the current date and whether the new privacy policy toast was clicked or closed.
+ *
+ * @param state - The application state containing the privacy policy data.
+ * @returns Boolean is True if the toast should be shown, and the number is the date the toast was last shown.
+ */
+export function getShowPrivacyPolicyToast(state: State): {
+  showPrivacyPolicyToast: boolean;
+  newPrivacyPolicyToastShownDate?: number;
+} {
+  const {
+    newPrivacyPolicyToastClickedOrClosed,
+    newPrivacyPolicyToastShownDate,
+    onboardingDate,
+  } = state.metamask || {};
+  const newPrivacyPolicyDate = new Date(PRIVACY_POLICY_DATE);
+  const currentDate = new Date(Date.now());
+
+  const showPrivacyPolicyToast =
+    !newPrivacyPolicyToastClickedOrClosed &&
+    currentDate >= newPrivacyPolicyDate &&
+    getIsPrivacyToastRecent(newPrivacyPolicyToastShownDate) &&
+    // users who onboarded before the privacy policy date should see the notice
+    // and
+    // old users who don't have onboardingDate set should see the notice
+    (!onboardingDate || onboardingDate < newPrivacyPolicyDate.valueOf());
+
+  return { showPrivacyPolicyToast, newPrivacyPolicyToastShownDate };
+}
+
+/**
+ * Returns true if the privacy policy toast was shown either never, or less than a day ago.
+ *
+ * @param newPrivacyPolicyToastShownDate
+ * @returns true if the privacy policy toast was shown either never, or less than a day ago
+ */
+function getIsPrivacyToastRecent(
+  newPrivacyPolicyToastShownDate?: number,
+): boolean {
+  if (!newPrivacyPolicyToastShownDate) {
+    return true;
+  }
+
+  const currentDate = new Date();
+  const oneDayInMilliseconds = 24 * 60 * 60 * 1000;
+  const newPrivacyPolicyToastShownDateObj = new Date(
+    newPrivacyPolicyToastShownDate,
+  );
+  const toastWasShownLessThanADayAgo =
+    currentDate.valueOf() - newPrivacyPolicyToastShownDateObj.valueOf() <
+    oneDayInMilliseconds;
+
+  return toastWasShownLessThanADayAgo;
+}
+
+export function setNewPrivacyPolicyToastShownDate(time: number) {
+  submitRequestToBackground('setNewPrivacyPolicyToastShownDate', [time]);
+}
+
+export function setNewPrivacyPolicyToastClickedOrClosed() {
+  submitRequestToBackground('setNewPrivacyPolicyToastClickedOrClosed');
+}
```

Comment:
I believe `submitRequestToBackground` would still need to be awaited/thenned even if it's no longer being called inside of a thunk callback.

```suggestion
export async function setNewPrivacyPolicyToastShownDate(time: number) {
  await submitRequestToBackground('setNewPrivacyPolicyToastShownDate', [time]);
}

export async function setNewPrivacyPolicyToastClickedOrClosed() {
  await submitRequestToBackground('setNewPrivacyPolicyToastClickedOrClosed');
}
```

Reply:
@Gudahtt I don't think there's any reason to wait for this to complete, but what do you think?

Reply:
My understanding is also that any requests to the background rely on the JSON-RPC API over the port streams and are therefore asynchronous, meaning we should `await` since `submitRequestToBackground` returns a promise.

I recall a contribution guideline that also recommends awaiting even when returning as it results in better stack traces in the event of errors?

Reply:
From that example, `a` is `setNewPrivacyPolicyToastShownDate`, `b` is `submitRequestToBackground`, and `c` is continuing on with `function PrivacyPolicyToast()`.  But `c` doesn't have to wait for `b` to finish.

Reply:
I think we might be conflating the following in this discussion:

```ts
b.then(() => c());
```
```ts
b.then(() => ...);
c();
```
In the second example, `b` does not synchronously block `c`.

With that in mind, how would you feel about at least removing the possibility of uncaught exceptions by adding error handling?
```ts
export function setNewPrivacyPolicyToastClickedOrClosed() {
  submitRequestToBackground('setNewPrivacyPolicyToastClickedOrClosed')
    .catch((error) => ...));
}
```

In general, floating promises are forbidden by our eslint config with the [`@typescript-eslint/no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises/) rule, which is also included in the recommended ruleset for `@typescript-eslint`. The only reason that error isn't showing up here is because extension is using eslint configs that are three years old. Adding a `.catch()` is a valid way to satisfy this lint rule.

Reply:
> a is setNewPrivacyPolicyToastShownDate, b is submitRequestToBackground, and c is continuing on with function PrivacyPolicyToast()

Having a bit of trouble understanding this. Wouldn't `PrivacyPolicyToast` be `a` since it's the outermost scope that the other two are nested in? Maybe the second half of `a` that follows `b` could be formulated as `c`, but then we don't seem to be examining the stack trace for an error in `c`? Not to mention `a` isn't an async function since it's a react component. 

This discussion probably belongs in a dedicated ticket in contributor-docs. I think having the test code available would be helpful as well.